File: backup/corriges/td/selection_activites/selection_activite.ml
(* CORRIGÉ PARTIEL DU TD : SÉLECTION D'ACTIVITÉ (PARTIE EN OCAML) *)
(* Par J. BENOUWT *)
(* Licence CC BY-NC-SA *)


(* --------------------------------------------------------- *)

type requete = {numero : int;
                debut : int;
                fin : int}

let prog_dynamique r =
    (* fonction auxiliaire qui détermine si les requêtes i et j sont compatibles *)
    let compatible i j =
        i.debut >= j.fin || j.debut >= i.fin
    in
    (* fonction auxiliaire qui renvoie la liste des requêtes compatibles avec i dans r *)
    let rec comp r i = match r with   
        | [] -> []
        | x::xs when compatible x i -> x::comp xs i
        | _::xs -> comp xs i
    in
    (* tableau associatif pour la mémoïsation *)
    let memo = Hashtbl.create 100 in
    (* fonction auxiliaire récursive avec mémoïsation *)
    let rec remplir r =
        (* on ne fait les calculs que s'ils n'ont pas déjà été faits et stockés *)
        begin if not (Hashtbl.mem memo r) then
            match r with
                | [] -> Hashtbl.add memo r 0 (* cas de base *)
                | x::xs -> Hashtbl.add memo r (max (remplir xs) (1 + remplir (comp xs x))) (* cas récursif *)
        end ;
        Hashtbl.find memo r
    (* on appelle la fonction avec l'ensemble de requêtes qui nous intéresse *)
    in remplir r

let test_exemple_du_sujet () =
    let r0 = {numero = 0; debut = 0; fin = 4} in
    let r1 = {numero = 1; debut = 5; fin = 10} in
    let r2 = {numero = 2; debut = 2; fin = 6} in
    let r3 = {numero = 3; debut = 8; fin = 10} in
    let r4 = {numero = 4; debut = 3; fin = 4} in
    let r5 = {numero = 5; debut = 5; fin = 7} in
    let r6 = {numero = 6; debut = 8; fin = 12} in
    let r7 = {numero = 7; debut = 3; fin = 7} in
    let r8 = {numero = 8; debut = 10; fin = 12} in
    let r = [r0; r1; r2; r3; r4; r5; r6; r7; r8] in
    print_string "Nombre de requêtes sélectionnées avec la programmation dynamique : " ;
    print_int (prog_dynamique r) ;
    print_endline "."

let _ = test_exemple_du_sujet ()

File: backup/corriges/td/selection_activites/selection_activite.c
/*
CORRIGÉ DU TD : SÉLECTION D'ACTIVITÉ (PARTIE EN C)
Par J. BENOUWT
Licence CC BY-NC-SA
*/


#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

/*---------------------------------------------------------*/

struct requete_s {
    int numero;
    int debut;
    int fin;
};
typedef struct requete_s requete;

// fonction auxiliaire qui trie le tableau R entre les indices d et f
void tri_aux(requete R[], int d, int f) {
    if (d < f) {
        requete pivot = R[f];
        int futur_indice_pivot = d;
        for (int i = d; i < f; i += 1) {
            if (R[i].fin < pivot.fin) { // on trie par rapport aux temps de fin
                requete tmp = R[i];
                R[i] = R[futur_indice_pivot];
                R[futur_indice_pivot] = tmp;
                futur_indice_pivot += 1;
            }    
        }
        R[f] = R[futur_indice_pivot];
        R[futur_indice_pivot] = pivot;
        tri_aux(R, d, futur_indice_pivot-1);
        tri_aux(R, futur_indice_pivot+1, f);
    }
}

void tri(int n, requete R[n]) {
    tri_aux(R, 0, n-1);
}

bool* glouton(int n, requete R[n]) {
    tri(n, R); // tri par ordre croissant des temps de fin
    bool* res = malloc(n * sizeof(bool));
    res[R[0].numero] = true; // on sélectionne l'activité qui se termine en premier
    int indice_derniere_requete_choisie = 0;
    for (int i = 1; i < n; i += 1) {
        if (R[i].debut >= R[indice_derniere_requete_choisie].fin) { // si la requête est compatible avec celle d'avant
            res[R[i].numero] = true; // on la sélectionne
            indice_derniere_requete_choisie = i;
        }
        else {
            res[R[i].numero] = false; // sinon on ne la sélectionne pas
        }
    }
    return res;
}

int main() {
    // exemple du sujet :
    requete r0 = {0, 0, 4};
    requete r1 = {1, 5, 10};
    requete r2 = {2, 2, 6};
    requete r3 = {3, 8, 10};
    requete r4 = {4, 3, 4};
    requete r5 = {5, 5, 7};
    requete r6 = {6, 8, 12};
    requete r7 = {7, 3, 7};
    requete r8 = {8, 10, 12};
    requete R[9] = {r0, r1, r2, r3, r4, r5, r6, r7, r8};
    int n = 9;
    bool* resultat = glouton(n, R);
    int nombre_requetes_selectionnees_par_glouton = 0;
    for (int i = 0; i < n; i += 1) {
        if (resultat[i]) {
            nombre_requetes_selectionnees_par_glouton += 1;
        }
    }
    printf("Nombre de requêtes sélectionnées par le glouton : %d.\n", nombre_requetes_selectionnees_par_glouton);
    free(resultat);
}